text/0000-associated-type-bounds.md

```diff
@@ -0,0 +1,117 @@
+- Feature Name: associated_type_bounds
+- Start Date: 2018-01-13
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Introduce the bound form `MyTrait<AssociatedType: Bounds>`, permitted anywhere
+a bound of the form `MyTrait<AssociatedType = T>` would be allowed. This form
+desugars to `MyTrait<AssociatedType = impl Bounds>`.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, when specifying a bound using a trait that has an associated
+type, the developer can specify the precise type via the syntax
+`MyTrait<AssociatedType = T>`. With the introduction of the `impl Trait`
+syntax for static-dispatch existential types, this syntax also permits
+`MyTrait<AssociatedType = impl Bounds>`, as a shorthand for introducing a
+new type variable and specifying those bounds.
+
+However, this introduces an unnecessary level of indirection that does not
+match the developer's intuition and mental model as well as it could. In
+particular, given the ability to write bounds on a type variable as `T: Bounds`,
+it makes sense to permit writing bounds on an associated type directly.
+This results in the simpler syntax `MyTrait<AssociatedType: Bounds>`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Instead of specifying a concrete type for an associated type, we can
+specify a bound on the associated type, to ensure that it implements
+specific traits, as seen in the example below:
+
+```rust
+fn print_all<T: Iterator<Item: Display>>(printables: T) {
+    for p in printables {
+        println!("{}", p);
+    }
+}
+```
+
+## In anonymous existential types
+
+```rust
+fn printables() -> impl Iterator<Item: Display> {
+    // ..
+}
+```
+
+## Further examples
+
+Instead of writing:
+
+```rust
+impl<I> Clone for Peekable<I>
+where
+    I: Clone + Iterator,
+    <I as Iterator>::Item: Clone,
+{
+    // ..
+}
+```
+
+you may write:
+
+```rust
+impl<I> Clone for Peekable<I>
+where
+    I: Clone + Iterator<Item: Clone>
+{
+    // ..
+}
+```
+
+or replace the `where` clause entirely:
+
+```rust
+impl<I: Clone + Iterator<Item: Clone>> Clone for Peekable<I> {
+    // ..
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The surface syntax `Trait<AssociatedType: Bounds>` should desugar to
+`Trait<AssociatedType = impl Bounds>` anywhere it appears. This syntax
```

Comment:
I'm not sure literally desugaring like this will be legal and will mean the same thing in all context.
Maybe it's better to use the desugaring from the examples above (`T: Trait<AssocTy: Bounds>` -> `T: Trait<AssocTy> + <T as Trait>::AssocTy: Bounds`) everywhere?

Reply:
Nice sugar, overall.

Reply:
So in other words: `T: Trait<AssocTy: Bounds>` desugars into adding `<T as Trait>::AssocTy: Bounds` to the list of `where` clauses? How do we deal with that in the case of the following?

```rust
fn printables() -> impl Iterator<Item: Display> {
    // ..
}
```

Reply:
@Centril `-> impl Iterator<Item = T>` is already desugared to two bounds, `X: Iterator` and `<X as Iterator>::Item == T`, where `X` is the `impl Iterator<Item = T>` *nominal* type.

Reply:
@petrochenkov You are more familiar with the inner workings of the compiler, so I think you can better specify the proper desugaring. Could you PR against our PR perhaps?

Reply:
Patch in: https://github.com/Centril/rfcs/pull/2